### PR TITLE
make GridTools::laplace_transform() to actually solve for displacement field

### PR DESCRIPTION
**Pull request by davydden**
Documentation of `GridTools::laplace_transform` suggested that one solves for displacement field (increment in vertex positions), but the actual implementation was solving for the total position, i.e. `x = X + u`. Sources for mesh motion i know always formulate this in terms of mesh displacement field (i updated documentation to reflect this). ~~But if there is a strong need, I can keep the original version by adding an extra boolean flag.~~ **edit**: done.

~~I mark it as WIP because i expected the tests without coefficient function to still pas (ie. `grid_transform_02`, `grid_transform_3d`), but the output files required changes.
I need to think about it more...~~

Below is the output from the new test:

grid_transform_03

### CODE REVIEW

**Comment by jppelteret** [comment]
@davydden I'm afraid that I can't look at this until Wednesday! Leaving for my travels in a couple of minutes.

**Comment by davydden** [comment]
On a second thought, minimizing gradient of position can also be used and essentially can be considered as mesh smoothing if no motion is defined by `new_points`.

**Comment by davydden** [comment]
I am satisfied with the current state of the PR, all `ctest -R "grid/grid_transform"` tests pass.

While modifying the function, we may as well add number of quadrature points as extra parameter and probably fe_degree.

p.s. Just noticed that the blessed output in `grid/grid_transform_02` does not look too good:

grid_transform_02 output

I can update the test to something more meaningful later.

**Comment by tjhei** [comment]
> Just noticed that the blessed output in grid/grid_transform_02 does not look too good:

Yeah, is this just a matter of making the displacement input smaller?

File: include/deal.II/grid/grid_tools.h [hunk 1]
```diff
@@ -295,12 +295,19 @@ namespace GridTools
    * typically, each of the vertices at the boundary of the triangulation is
    * mapped to the corresponding points in the @p new_points map.
    *
-   * The way this function works is that it solves a Laplace equation for each
+   * The unknown displacement field $u_d(\mathbf x)$ in direction $d$ is obtained from
+   * the minimization problem
+   * \f[
+   *   \rm{min}\, \frac{1}{2}\int c(x) \mathbf \nabla u_d(\mathbf x) \cdot \mathbf \nabla u_d(\mathbf x) \,\rm d v
+   * \f]
+   * subject to prescribed constraints. The minizer is obtained by solving the Laplace equation
```
Comment (tjhei) [inline include/deal.II/grid/grid_tools.h]: "minimizer"?

**Comment by davydden** [comment]
> Yeah, is this just a matter of making the displacement input smaller?

correct, trivial to fix. Just don't want to touch it before the current part is finalized.

**Comment by davydden** [comment]
> What is the reason for keeping the old variant solving for positions, is that just for backwards compatibility?

yes. I actually think it was a bug, but did not want to remove it completely, just in case someone is using.

> I would deprecate the old and create a new function with different arguments or name.

that's also possible. I can shift implementation into the internal function and use it as-is in depricated variant and in the new one, which, can have a few extra arguments like `n_q_points` and `degree`.

Let's see what others have to say...

**Comment by davydden** [comment]
@tjhei :
I would actually keep the old variant. Maybe someone wants to use it as a mesh smoothing where you start from non-uniform (distorted) mesh and make it more uniform by requiring norm of gradient of position field to be minimal (easy to imagine in 1D).

@bangerth :  
since you recently revised this function https://github.com/dealii/dealii/commit/f2552d7f12e7ecd443f3bb9e669ede121a1d2244 and also commited the patch from myself and @jppelteret in 2014 https://github.com/dealii/dealii/commit/c6269079be0c852190f1924a24063dd7016d9cf1, i added you as a reviewer to this PR.

**Comment by luca-heltai** [comment]
I'd keep the old variant. MappingFE uses *absolute* position for the displacement vector (for a general FE it is not obvious what a relative displacement should be...). In that case, this is the natural function to use for mesh deformation.

**Comment by davydden** [comment]
@luca-heltai are u ok with keeping the new variant on by default?

> for a general FE it is not obvious what a relative displacement should be...

At least in the context of mesh motion for vertices, one should be able to use any continuous FE space. it may not work well for mesh motion, but it's perfectly ok from minimization perspective subject to some point constraints.

**Comment by davydden** [comment]
P.s. The problem with the old variant is that it may still deform the mesh even if displacements specified are zero (i.e. New positions coincide with old). It is irrelevant to this PR, but I don't agree that minimizing for absolute position is more natural.

**Comment by davydden** [comment]
@luca-heltai i expanded the discussion on the difference between the two approaches.

**Comment by davydden** [comment]
i changed `grid_transform_02` to look better:
grid_transform_02
The PR should be ready for review.

File: include/deal.II/grid/grid_tools.h [hunk 2]
```diff
@@ -295,12 +295,19 @@ namespace GridTools
...
-   * The way this function works is that it solves a Laplace equation for each
+   * The unknown displacement field $u_d(\mathbf x)$ in direction $d$ is obtained from
+   * the minimization problem
+   * \f[
+   *   \rm{min}\, \frac{1}{2}\int c(\mathbf x) \mathbf \nabla u_d(\mathbf x) \cdot \mathbf \nabla u_d(\mathbf x) \,\rm d v
+   * \f]
+   * subject to prescribed constraints. The minimizer is obtained by solving the Laplace equation
    * of the dim components of a displacement field that maps the current
-   * domain into one described by @p new_points . The @p new_points array
-   * therefore represents the boundary values of this displacement field. The
```
Comment (davydden) [inline include/deal.II/grid/grid_tools.h]: here and below one can clearly see that the documentation implied the solution for the displacement field as opposed to the total/final position of vertices.

**Comment by bangerth** [comment]
I'm going to look through this later, but I did want to add the following piece of information:

Let's say you solve for displacement `u`, then you would solve the equation
```
  -Delta  u  = 0,
  u = 0    on the boundary
```
On the other hand, if you solve for the position `U = X + u`, then you would solve for the equation
```
  -Delta  U  = 0,
  U = X    on the boundary
```
But this will yield the same solution, because `- Delta X = 0`. So as far as the result is concerned, there is really no difference between the two formulations -- you just add a harmonic function to what you call the solution, but the vertex displacement you will get will be exactly the same between the two formulations.

**Comment by davydden** [comment]
@bangert I think you are mistaken on this one. Specifically -Delta X != 0. Think of the 1D mesh which is denser towards one end. If u prescribe zero displacements on the ends, the minimum of displacement field subject to constraints is zero. Whereas if u minimize for gradient of total position, subject to prescribed values, then this may lead to redistribution of internal nodes in order to make the gradient of position field minimal over the domain.

**Comment by bangerth** [comment]
No, no, `-Delta X` really is zero :-) That's because `grad X = \frac{\partial x_i}{\partial x_j} = I` (the identity matrix), so the second derivatives are zero.

**Comment by davydden** [comment]
I don't think so ;-) Let's say new position is x=X+u. But the Laplace is solved in **reference coordinates** (current mesh position, X). So the field u prescribe, i.e. x on \patial \Onega does not have to have zero gradient wrt X. It may have, though, in some simple case. Ps. Typing from mob, sorry for sloppy indentation.

**Comment by davydden** [comment]
I will give some numerical examples tomorrow...

File: include/deal.II/grid/grid_tools.h [hunk 3]
```diff
@@ -295,12 +295,19 @@ namespace GridTools
    * typically, each of the vertices at the boundary of the triangulation is
    * mapped to the corresponding points in the @p new_points map.
    *
-   * The way this function works is that it solves a Laplace equation for each
+   * The unknown displacement field $u_d(\mathbf x)$ in direction $d$ is obtained from
+   * the minimization problem
+   * \f[
+   *   \rm{min}\, \frac{1}{2}\int c(\mathbf x) \mathbf \nabla u_d(\mathbf x) \cdot \mathbf \nabla u_d(\mathbf x) \,\rm d v
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Why not use `\min` directly?

Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Can you break this line so that I get to see everything in the github patch review page? :-)

File: include/deal.II/grid/grid_tools.h [hunk 4]
```diff
@@ -295,12 +295,19 @@ namespace GridTools
...
+   * \f]
+   * subject to prescribed constraints. The minimizer is obtained by solving the Laplace equation
    * of the dim components of a displacement field that maps the current
-   * domain into one described by @p new_points . The @p new_points array
-   * therefore represents the boundary values of this displacement field. The
-   * function then evaluates this displacement field at each vertex in the
-   * interior and uses it to place the mapped vertex where the displacement
+   * domain into one described by @p new_points . Linear finite elements with
+   * the fourth order Gaussian quadrature are used. The difference between the
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: -> with fourth order ...

But it's not clear to me what fourth order actually refers to. Do you mean 4 quadrature points in each coordinate direction?

File: include/deal.II/grid/grid_tools.h [hunk 5]
```diff
@@ -295,12 +295,19 @@ namespace GridTools
...
+   * vertex positions specified in @p new_points and their current value in @p tria
+   * therefore represents the prescribed values of this displacement field. The
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: ...at the boundary of the domain, or more precisely at all of those locations for which @p new_points provides values (which may be at part of the boundary, or even in the interior of the domain)

File: include/deal.II/grid/grid_tools.h [hunk 6]
```diff
@@ -323,12 +330,21 @@ namespace GridTools
    * Should this function be provided, sensible results can only be expected
    * if all coefficients are positive.
    *
+   * @param[in] solve_for_position_field If set to <code>true</code>, the
+   * minimization problem is formulated with respect to the final vertex positions
+   * as opposed to their displacement. Note that this case may lead to non-zero
+   * displacement of un-constrained vertices even when the @p new_points
+   * don't specify any deformation. That is because the minimum corresponds to
+   * smooth vertex position field, instead if their displacement. Thereby this
+   * case can be considered as a variation of Laplacian smoothing technique.
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: I don't think I believe this is correct, for the reason we're discussing in the general discussion section of this PR.

**Comment by bangerth** [comment]
I still think I'm right :-)

What this function does is the following: On the current domain (with variables x), it solves the problem
```
  -Delta u = 0
  u = newpoint(x)-x    on the boundary
```
Let's assume that the new points cover all boundary vertices, then the boundary values of `u` are all zero, and consequently the solution of the PDE is zero. Since `u` is the position update for each vertex in the interior, this means that no nodes are moved.

Alternatively, one can write things in terms of the new positions `U`. Then we solve
```
  -Delta U = 0
  U = newpoint(x)    on the boundary
```
and assign (instead of add to) the locations `U` to the vertices. Now, if the new vertex locations equal the old vertex locations, then `newpoint(x)=x`. So let me write `U=V+x` and see what equation `V` solves:
```
  -Delta (V+x) = 0
  V+x = x    on the boundary
```
But because `-Delta x = div (grad x) = div I = 0`, this means that `V` has to solve
```
  -Delta V = 0
  V = 0    on the boundary
```
which is again `V=0`, so the new vertex positions `U(x)=V(x)+x=x` equal the old positions `x`.

In other words, at least for the case where you pose newpoint locations all the way around the boundary, the two formulations are exactly the same. I do think that's also true for cases where you have them around only part of the boundary -- if the new locations are the same as the old locations, neither formulation yields any displacement.

**Comment by davydden** [comment]
@bangert ok, i think you are both right and wrong ;-) right in case of the homogeneous Laplace, wrong when a heterogeneity coefficient (diffusivity) is prescribed to make some cells more rigid. The whole reason why I differ into this function is that I saw huge descripancy between my own implementation and the one in dealii.

**Comment by bangerth** [comment]
Ah, yes, I can believe that to be true. I think you only want to use a coefficient if you use the displacement, not the position formulation.

**Comment by davydden** [comment]
> I think you only want to use a coefficient if you use the displacement, not the position formulation.

Correct. I have not seen anyone using coefficient with position field. I will update the function description to reflect this.

**Comment by davydden** [comment]
@bangerth i update the PR according to your comments in the discussion.

**Comment by davydden** [comment]
@bangerth ping.

File: include/deal.II/grid/grid_tools.h [hunk 7]
```diff
@@ -323,12 +338,21 @@ namespace GridTools
...
+   * value of this parameter is <code>false</code>.
+   *
    * @note This function is not currently implemented for the 1d case.
    */
   template <int dim>
   void laplace_transform (const std::map<unsigned int,Point<dim> > &new_points,
                           Triangulation<dim> &tria,
-                          const Function<dim,double> *coefficient = 0);
+                          const Function<dim,double> *coefficient = 0,
+                          const bool solve_for_position_field = false);
```
Comment (tjhei) [inline include/deal.II/grid/grid_tools.h]: Can we rename it to something like ``solve_for_absolute_positions`` or ``solve_for_disclacements``?

File: include/deal.II/grid/grid_tools.h [hunk 8]
```diff
@@ -323,12 +338,21 @@ namespace GridTools
    * Should this function be provided, sensible results can only be expected
    * if all coefficients are positive.
    *
+   * @param[in] solve_for_position_field If set to <code>true</code>, the
+   * minimization problem is formulated with respect to the final vertex positions
+   * as opposed to their displacement. The two formulations are equivalent for
+   * the homogeneous problem (default value of @p coefficient), but they
+   * result in very different mesh motion otherwise. Since in most cases one will
+   * be using heterogeneity coefficient in displacement formulation, the default
```
Comment (tjhei) [inline include/deal.II/grid/grid_tools.h]: "a non-constant coefficient"?

File: source/grid/grid_tools.cc [hunk 9]
```diff
@@ -693,7 +693,8 @@ namespace GridTools
   template <>
   void laplace_transform (const std::map<unsigned int,Point<1> > &,
                           Triangulation<1> &,
-                          const Function<1> *)
+                          const Function<1> *,
+                          const bool solve_for_position_field)
```
Comment (tjhei) [inline source/grid/grid_tools.cc]: remove the name of the variable here

File: tests/grid/grid_transform_02.cc [hunk 10]
```diff
@@ -83,7 +83,7 @@ int main ()
             }
 
   // then compute new point locations and output the result
-  GridTools::laplace_transform  (new_points,tria);
+  GridTools::laplace_transform  (new_points,tria, static_cast<const Function<dim, double> *>(NULL), true);
```
Comment (tjhei) [inline tests/grid/grid_transform_02.cc]: is this cast necessary? If yes, add a comment.

Comment (davydden) [inline tests/grid/grid_transform_02.cc]: it is to call the old code.

Comment (davydden) [inline tests/grid/grid_transform_02.cc]: oh, i misread, yes, i could not compile otherwise.

Comment (tjhei) [inline tests/grid/grid_transform_02.cc]: because the compiler can not deduce the template argument?

File: doc/news/changes/minor/20170217DenisDavydov [hunk 11]
```diff
@@ -0,0 +1,6 @@
+Fixed: Fixed a bug in GridTools::laplace_transform() which was previously solving
+for the final position field instead of displacement. The two formulations are
+not equivalent when heterogeneity coefficient is provided, which is commonly
```
Comment (jppelteret) [inline doc/news/changes/minor/20170217DenisDavydov]: when a heterogeneity coefficient

File: include/deal.II/grid/grid_tools.h [hunk 12]
```diff
@@ -295,12 +295,27 @@ namespace GridTools
...
+   *   \,\rm d v
+   * \f]
+   * subject to prescribed constraints. The minimizer is obtained by solving the Laplace equation
    * of the dim components of a displacement field that maps the current
-   * domain into one described by @p new_points . The @p new_points array
-   * therefore represents the boundary values of this displacement field. The
-   * function then evaluates this displacement field at each vertex in the
-   * interior and uses it to place the mapped vertex where the displacement
+   * domain into one described by @p new_points . Linear finite elements with
+   * four Gaussian quadrature point in each direction are used. The difference between the
```
Comment (jppelteret) [inline include/deal.II/grid/grid_tools.h]: quadrature points

File: include/deal.II/grid/grid_tools.h [hunk 13]
```diff
@@ -295,12 +295,27 @@ namespace GridTools
...
+   * vertex positions specified in @p new_points and their current value in @p tria
+   * therefore represents the prescribed values of this displacement field at
+   * the boundary of the domain, or more precisely at all of those locations for
+   * which @p new_points provides values (which may be at part of the boundary,
+   * or even in the interior of the domain). The
```
Comment (jppelteret) [inline include/deal.II/grid/grid_tools.h]: Strange line break?

File: include/deal.II/grid/grid_tools.h [hunk 14]
```diff
@@ -295,12 +295,27 @@ namespace GridTools
...
+   * function then evaluates this displacement field at each un-constrained
```
Comment (jppelteret) [inline include/deal.II/grid/grid_tools.h]: unconstrained

File: tests/grid/grid_transform_03.cc [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+  Triangulation< dim >  tria;
+  std::map< unsigned int, Point< dim > > new_points;
+  const unsigned int N = 8;
+  GridGenerator::subdivided_hyper_cube (tria, N, -5, 5);
+
+  // find the vertex at the origin
+  Triangulation<dim>::active_cell_iterator
+  cell = GridTools::find_active_cell_around_point (tria,Point<dim>());
+
+  unsigned int best_vertex   = cell->vertex_index(0);//vertex number on local triangulation
```
Comment (jppelteret) [inline tests/grid/grid_transform_03.cc]: Petandic comment: Space required between semi-colon and comment slashes

File: tests/grid/grid_transform_02.cc (hunk 10 above)
Comment (jppelteret) [inline tests/grid/grid_transform_02.cc]: I believe its because the default value for the last argument is `false`, and its set to `true` here.

Comment (tjhei) [inline tests/grid/grid_transform_02.cc]: Sure, but that doesn't explain why you need to add the cast for the third argument or am I missing something?

Comment (davydden) [inline tests/grid/grid_transform_02.cc]: i would also expect this cast not to be needed, but
```
236: /Users/davydden/libs-sources/deal.ii/davydden/tests/grid/grid_transform_02.cc:86:3: error: no matching function for call to 'laplace_transform'
236:   GridTools::laplace_transform  (new_points,tria, NULL, true);
236:   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
236: /Users/davydden/libs-sources/deal.ii/davydden/include/deal.II/grid/grid_tools.h:352:8: note: candidate template ignored: could not match 'const Function<dim, double> *' against 'long'
236:   void laplace_transform (const std::map<unsigned int,Point<dim> > &new_points,
236:        ^
236: 1 error generated.
```

Comment (davydden) [inline tests/grid/grid_transform_02.cc]: maybe it's related to the fact that  the pointer is to templated class `Function<dim>`.

Comment (davydden) [inline tests/grid/grid_transform_02.cc]: > because the compiler can not deduce the template argument?

most likely. I will try to explicitly specify template argument...

Comment (tjhei) [inline tests/grid/grid_transform_02.cc]: No, this is because NULL is not a function pointer but basically a (long) int and so template deduction doesn't kick in. Learned something new today. :-)

**Comment by davydden** [comment]
should be ready:
```
$ ctest -R "grid_transform"
Test project /Users/davydden/libs-sources/deal.ii/builds/spack
      Start 233: grid/grid_transform.debug
 1/10 Test #233: grid/grid_transform.debug .................   Passed   14.03 sec
      Start 234: grid/grid_transform.release
 2/10 Test #234: grid/grid_transform.release ...............   Passed   10.57 sec
      Start 235: grid/grid_transform_02.debug
 3/10 Test #235: grid/grid_transform_02.debug ..............   Passed    1.26 sec
      Start 236: grid/grid_transform_02.release
 4/10 Test #236: grid/grid_transform_02.release ............   Passed    1.35 sec
      Start 237: grid/grid_transform_03.debug
 5/10 Test #237: grid/grid_transform_03.debug ..............   Passed   11.55 sec
      Start 238: grid/grid_transform_03.release
 6/10 Test #238: grid/grid_transform_03.release ............   Passed   10.49 sec
      Start 239: grid/grid_transform_3d.debug
 7/10 Test #239: grid/grid_transform_3d.debug ..............   Passed    1.15 sec
      Start 240: grid/grid_transform_3d.release
 8/10 Test #240: grid/grid_transform_3d.release ............   Passed    1.24 sec
      Start 241: grid/grid_transform_coefficient.debug
 9/10 Test #241: grid/grid_transform_coefficient.debug .....   Passed   11.54 sec
      Start 242: grid/grid_transform_coefficient.release
10/10 Test #242: grid/grid_transform_coefficient.release ...   Passed   10.77 sec
```

File: doc/news/changes/minor/20170217DenisDavydov [hunk 16]
```diff
@@ -0,0 +1,6 @@
+Fixed: Fixed a bug in GridTools::laplace_transform() which was previously solving
+for the final position field instead of displacement. The two formulations are
+not equivalent when a heterogeneity coefficient is provided, which is commonly
```
Comment (bangerth) [inline doc/news/changes/minor/20170217DenisDavydov]: say "non-constant coefficient" -- I think that's easier to understand.

File: include/deal.II/grid/grid_tools.h [hunk 17]
```diff
@@ -295,12 +295,27 @@ namespace GridTools
...
-   * The way this function works is that it solves a Laplace equation for each
+   * The unknown displacement field $u_d(\mathbf x)$ in direction $d$ is obtained from
+   * the minimization problem
+   * \f[
+   *   \min\, \int
+   *   \frac{1}{2}
+   *   c(\mathbf x)
+   *   \mathbf \nabla u_d(\mathbf x) \cdot
+   *   \mathbf \nabla u_d(\mathbf x)
+   *   \,\rm d v
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Since you use `x` as the variable, maybe also use `dx`? Or `dv(x)` if you want?
